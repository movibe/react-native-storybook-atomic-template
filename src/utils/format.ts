export const formatPoints = (points: string | number): string => {
  const minus = Number(points) < 0

  let n = Number(points).toString()
  let r = ''
  let x = 0

  for (let i = n.length; i > 0; i--) {
    r += n.substr(i - 1, 1) + (x == 2 && i != 1 ? '.' : '')
    x = x == 2 ? 0 : x + 1
  }
  return `${minus ? '-' : ''}${r.split('').reverse().join('')}`
}

export interface IFlightsStops {
  name: string
  iata: string
  duration: string
}
