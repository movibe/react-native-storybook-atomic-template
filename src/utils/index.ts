export * from './faker'
export * from './qualityAssurance'
export * from './format'
export * from './withTheme'
