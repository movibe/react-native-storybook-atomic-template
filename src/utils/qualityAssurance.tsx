import { Platform } from 'react-native'

interface ItestID {
  accessible?: boolean
  testID: string
  accessibilityLabel?: string
}
const testID = (testID: string): ItestID =>
  Platform.OS === 'android'
    ? {
        testID,
        accessible: true,
        accessibilityLabel: testID,
      }
    : {
        testID,
        accessible: false,
      }

type componentType =
  | 'ActivityIndicator'
  | 'View'
  | 'Lottie'
  | 'ScrollView'
  | 'Modal'
  | 'TextInput'
  | 'Image'
  | 'Text'
  | 'ImageBackground'
  | 'TouchableOpacity'
  | 'Radio'
  | 'Checkbox'

export const assignTestId = (componentType: componentType, componentId: string): ItestID => {
  const id = `${componentType}_${componentId}`
  return testID(id)
}
export const mockTestID = (componentType: componentType, componentId: string): ItestID => assignTestId(componentType, componentId)
