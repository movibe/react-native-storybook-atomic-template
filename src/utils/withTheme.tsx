import { boolean, radios } from '@storybook/addon-knobs'
import * as React from 'react'
import styled, { DefaultTheme, ThemeProvider } from 'styled-components/native'

import { Column } from '../atomic/atoms/Flex'
import { themes } from '../tokens'

const options = {
  Light: 'light',
  Dark: 'dark',
}

interface IProps {
  children: any
  padding?: boolean
}

const Background = styled.View`
  background-color: ${({ theme }) => theme.colors.background};
  height: 100vh;
  flex: 1;
`
export const StorybookHoc: React.FC<IProps> = ({ children, padding = true }) => {
  const color = radios('theme_color', options, 'light')
  const _theme: DefaultTheme = color === 'dark' ? themes.dark : themes.light

  return (
    <ThemeProvider theme={_theme}>
      <Background>
        <Column content={boolean('theme_padding', !!padding)} flex={1}>
          {children}
        </Column>
      </Background>
    </ThemeProvider>
  )
}

export const withTheme = (fn: any) => <StorybookHoc>{fn()}</StorybookHoc>
