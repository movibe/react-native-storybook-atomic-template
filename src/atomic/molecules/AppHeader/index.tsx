import * as React from 'react'
import { TouchableOpacity } from 'react-native'

import logoImage from '../../../assets/images/logo.svg'
import { Column } from '../../../atomic/atoms/Flex'
import { assignTestId } from '../../../utils'
import * as S from './styles'

export interface IAppHeaderProps {
  testID?: string
  title?: string
  subtitle?: string
  logo?: boolean
  onBack?: () => void
  onFilter?: () => void
}

export const AppHeader: React.FC<IAppHeaderProps> = ({ testID = 'AppHeader', title, subtitle, logo, onBack, onFilter }): JSX.Element => {
  return (
    <S.Container testID={testID}>
      <S.ContainerTop alignItems='center'>
        <S.Left>
          {onBack && (
            <TouchableOpacity onPress={onBack} {...assignTestId('TouchableOpacity', `${testID}_onBack`)}>
              <S.ArrowBackIcon />
            </TouchableOpacity>
          )}
        </S.Left>
        <S.Body>
          {title && <S.Title testID={`${testID}_title`}>{title}</S.Title>}
          {logo && <S.Logo testID={`${testID}_logo`} source={logoImage} />}
        </S.Body>
        <S.Right></S.Right>
      </S.ContainerTop>
      {subtitle && (
        <S.ContainerSubtitle>
          <Column flex={8} justifyContent='center'>
            {subtitle && <S.SubTitle testID={`${testID}_subtitle`}>{subtitle}</S.SubTitle>}
          </Column>
          <Column flex={1} justifyContent='center' alignItems='flex-end'>
            {onFilter && (
              <TouchableOpacity onPress={onFilter} {...assignTestId('TouchableOpacity', `${testID}_onFilter`)}>
                <S.FilterIcon />
              </TouchableOpacity>
            )}
          </Column>
        </S.ContainerSubtitle>
      )}
    </S.Container>
  )
}
