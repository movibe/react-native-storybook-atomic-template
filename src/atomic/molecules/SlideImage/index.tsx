import React, { useState } from 'react'
import { ImageSourcePropType, View } from 'react-native'
import Carousel, { Pagination } from 'react-native-snap-carousel'
import styled from 'styled-components/native'

import { Column } from '../../../atomic/atoms/Flex'
import { Image } from '../../../atomic/atoms/Image'
import { assignTestId } from '../../../utils'

export interface ISlideImageProps {
  testID?: string
  images: Array<ImageSourcePropType>
  width: number
  height?: number
  onPress?: () => void
  autoPlay?: boolean
  firstItem?: number
  showPagination?: boolean
}

const Container = styled(Column)`
  background-color: ${({ theme }) => theme.colors.background};
  border-bottom-width: 1px;
  border-color: ${({ theme }) => theme.colors.disabled};
  width: 100%;
`
export const SlideImage: React.FC<ISlideImageProps> = ({
  testID = 'SlideImage',
  images,
  width,
  onPress,
  height,
  firstItem,
  autoPlay,
  showPagination,
}): JSX.Element => {
  const [activeSlide, setActiveSlide] = useState(firstItem || 0)

  const renderItem = ({ item, index }) => (
    <Column testID={`${testID}_slide_item`} onPress={onPress} key={index}>
      <Image testID={`${testID}_slide_item_image`} source={{ ...item, height: height, width: width }} />
    </Column>
  )

  const onChangeSlide = (index) => {
    setActiveSlide(index)
  }

  return (
    <Container alignItems='center' testID={testID}>
      <Carousel
        {...assignTestId('View', `${testID}_slide`)}
        data={images}
        renderItem={renderItem}
        sliderWidth={width}
        itemWidth={width}
        firstItem={firstItem}
        onSnapToItem={onChangeSlide}
        autoplay={autoPlay}
        activeSlideAlignment='center'
      />
      {showPagination && (
        <View {...assignTestId('View', `SlideImage_pagination`)}>
          <Pagination
            dotsLength={images.length}
            activeDotIndex={activeSlide}
            dotStyle={{
              width: 10,
              height: 10,
              borderRadius: 5,
              backgroundColor: '#828282',
            }}
            inactiveDotStyle={{
              backgroundColor: '#D8D8D8',
            }}
            inactiveDotOpacity={0.4}
            inactiveDotScale={0.6}
          />
        </View>
      )}
    </Container>
  )
}
