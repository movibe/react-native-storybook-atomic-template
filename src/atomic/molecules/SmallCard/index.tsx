import * as React from 'react'
import { ImageSourcePropType, TouchableOpacityProps } from 'react-native'

import { Text as NText } from '../../../atomic/atoms/Text'
import styled from 'styled-components/native'
import { assignTestId } from '../../../utils'

const Container = styled.TouchableOpacity`
  box-shadow: ${(props) => props.theme.boxShadow};
  border-radius: ${(props) => props.theme.borderRadius};
  overflow: hidden;
`

const Background = styled.ImageBackground`
  flex: 1;
  overflow: hidden;
  justify-content: flex-end;
`

const Overlay = styled.View`
  flex: 0.5;
  justify-content: flex-end;
  background-color: rgba(0, 0, 0, 0.5);
`

const Text = styled(NText)`
  color: #fff;
  font-size: 14px;
  font-weight: bold;
  margin: 10px;
`

export interface ISmallCardProps extends TouchableOpacityProps {
  testID?: string
  text?: string
  image: ImageSourcePropType
  gradient?: boolean
  onPress?: () => void
  width?: number
  height?: number
}

export const SmallCard: React.FC<ISmallCardProps> = ({
  testID = 'SmallCard',
  text,
  image,
  gradient,
  width = 120,
  height = 126,
  ...props
}): JSX.Element => {
  return (
    <Container {...assignTestId('TouchableOpacity', testID)} {...props} style={{ width, height }}>
      <Background source={image}>
        {gradient ? (
          <Overlay>
            <Text testID={testID}>{text}</Text>
          </Overlay>
        ) : (
          <Text testID={testID}>{text}</Text>
        )}
      </Background>
    </Container>
  )
}
