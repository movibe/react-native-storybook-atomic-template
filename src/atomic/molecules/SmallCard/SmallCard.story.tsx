import { action } from '@storybook/addon-actions'
import { boolean, text } from '@storybook/addon-knobs'
import * as React from 'react'

import { SmallCard } from '../../../atomic/molecules/SmallCard'
import { faker } from '../../../utils'

export default {
  title: 'Molecules/SmallCard',
}

export const normal = () => (
  <SmallCard
    onPress={action('onPress')}
    text={text('text', 'Descontos em combustÃ­vel')}
    gradient={boolean('gradient', false)}
    image={{
      uri: faker.image.fashion(),
    }}
  />
)
