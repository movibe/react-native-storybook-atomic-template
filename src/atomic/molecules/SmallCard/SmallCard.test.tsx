import * as React from 'react'
import { fireEvent, render } from 'react-native-testing-library'

import { ISmallCardProps, SmallCard } from '../../../atomic/molecules/SmallCard'
import { mockTestID, StorybookHoc as wrapper } from '../../../utils'

describe('Molecules/SmallCard', () => {
  let props: ISmallCardProps = {
    testID: 'SmallCard',
    image: {
      uri: 'https://i.redd.it/mjauu6mwms6z.jpg',
    },
  }

  const { toJSON, findByTestId, update } = render(<SmallCard {...props} />, {
    wrapper,
  })

  const el_container = mockTestID('TouchableOpacity', props.testID!).testID
  const el_container_press = mockTestID('TouchableOpacity', props.testID!).testID
  const el_text = mockTestID('Text', props.testID!).testID

  it(`render component #${el_container}`, async () => {
    const sut = await findByTestId(el_container)
    expect(sut).toBeTruthy()
  })

  it(`render component #${el_text}`, async () => {
    props.text = 'Exemplo'
    update(<SmallCard {...props} />)
    const sut = await findByTestId(el_text)
    expect(sut).toBeTruthy()
    expect(sut.children).toEqual([props.text])
  })
  it(`render component #${el_container_press}`, async () => {
    props.onPress = jest.fn()
    update(<SmallCard {...props} />)
    const sut = await findByTestId(el_container_press)
    fireEvent(sut, 'press')
    expect(props.onPress).toBeCalledTimes(1)
  })

  it('should render snapshot', () => {
    expect(toJSON()).toMatchSnapshot()
  })
})
