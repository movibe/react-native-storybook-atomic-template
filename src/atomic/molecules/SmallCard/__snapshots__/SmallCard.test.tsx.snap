// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/SmallCard should render snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "height": "100vh",
      },
    ]
  }
>
  <View
    accessible={false}
    style={
      Array [
        undefined,
        Object {
          "alignItems": undefined,
          "borderColor": undefined,
          "borderWidth": 0,
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": undefined,
          "padding": 10,
        },
      ]
    }
    testID="View_View"
  >
    <View
      accessible={false}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderRadius": 10,
          "height": 126,
          "opacity": 1,
          "overflow": "hidden",
          "shadowColor": "rgba(0,0,0,0.2)",
          "shadowOffset": Object {
            "height": 5,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 15,
          "width": 120,
        }
      }
      testID="TouchableOpacity_SmallCard"
    >
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "flex-end",
              "overflow": "hidden",
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "uri": "https://i.redd.it/mjauu6mwms6z.jpg",
            }
          }
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "height": undefined,
                "width": undefined,
              },
              undefined,
            ]
          }
        />
        <Text
          accessible={false}
          style={
            Array [
              Object {
                "color": "#444",
                "fontSize": 12,
                "textTransform": "none",
              },
              Object {
                "color": "#fff",
                "fontSize": 14,
                "fontWeight": "bold",
                "marginBottom": 10,
                "marginLeft": 10,
                "marginRight": 10,
                "marginTop": 10,
              },
            ]
          }
          testID="Text_SmallCard"
        >
          Exemplo
        </Text>
      </View>
    </View>
  </View>
</View>
`;
